ec/ecresource/deploymentresource/apmstate/apm_expanders.go

```diff
@@ -106,8 +116,81 @@ func expandTopology(raw interface{}) ([]*models.ApmTopologyElement, error) {
 			elem.ZoneCount = int32(zones.(int))
 		}
 
+		if c, ok := topology["config"]; ok {
+			elem.Apm = expandConfig(c)
+		}
+
 		res = append(res, &elem)
 	}
 
 	return res, nil
 }
+
+func expandConfig(raw interface{}) *models.ApmConfiguration {
+	var res = new(models.ApmConfiguration)
+	for _, rawCfg := range raw.([]interface{}) {
+		var cfg = rawCfg.(map[string]interface{})
+		if dockerImage, ok := cfg["docker_image"]; ok {
+			res.DockerImage = dockerImage.(string)
+		}
+
+		if debugEnabled, ok := cfg["debug_enabled"]; ok {
+			initSystemSettings(res)
+			res.SystemSettings.DebugEnabled = ec.Bool(debugEnabled.(bool))
+		}
+
+		if pass, ok := cfg["elasticsearch_password"]; ok {
+			initSystemSettings(res)
+			res.SystemSettings.ElasticsearchPassword = pass.(string)
+		}
+
+		if u, ok := cfg["elasticsearch_url"]; ok {
+			initSystemSettings(res)
+			res.SystemSettings.ElasticsearchURL = u.(string)
+		}
+
+		if user, ok := cfg["elasticsearch_username"]; ok {
+			initSystemSettings(res)
+			res.SystemSettings.ElasticsearchUsername = user.(string)
+		}
+
+		if u, ok := cfg["kibana_url"]; ok {
+			initSystemSettings(res)
+			res.SystemSettings.KibanaURL = u.(string)
+		}
+
+		if token, ok := cfg["secret_token"]; ok {
+			initSystemSettings(res)
+			res.SystemSettings.SecretToken = token.(string)
+		}
+
+		if settings, ok := cfg["user_settings_json"]; ok && settings != nil {
+			if s, ok := settings.(string); ok && s != "" {
+				res.UserSettingsJSON = settings
+			}
+		}
+		if settings, ok := cfg["user_settings_override_json"]; ok && settings != nil {
+			if s, ok := settings.(string); ok && s != "" {
+				res.UserSettingsOverrideJSON = settings
+			}
+		}
+		if settings, ok := cfg["user_settings_yaml"]; ok {
+			res.UserSettingsYaml = settings.(string)
+		}
+		if settings, ok := cfg["user_settings_override_yaml"]; ok {
+			res.UserSettingsOverrideYaml = settings.(string)
+		}
+	}
+
+	if !reflect.DeepEqual(res, new(models.ApmConfiguration)) {
+		return res
+	}
+
+	return nil
+}
+
+func initSystemSettings(cfg *models.ApmConfiguration) {
```

Comment:
Why does this function need to be called so many times within `expandConfig` ?

Reply:
I'll address this, it isn't necessary anymore

Reply:
b244157 changes the behaviour to match the defaults in the schema, there's no case where the `SystemSettings` would be empty now, so it doesn't make sense to add the complexity to tackle that